qiskit_experiments/data_processing/nodes.py

```diff
@@ -0,0 +1,141 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Different data analysis steps."""
+
+from abc import abstractmethod
+from typing import Any, Dict, Optional, Tuple
+import numpy as np
+
+from qiskit_experiments.data_processing.base import DataAction
+
+
+class IQPart(DataAction):
+    """Abstract class for IQ data post-processing."""
+
+    def __init__(self, scale: Optional[float] = 1.0, average: bool = False):
+        """
+        Args:
+            scale: scale by which to multiply the real part of the data.
+            average: if True the single-shots are averaged.
+        """
+        self.scale = scale
+        self.average = average
+        super().__init__()
+        self._accepted_inputs = ["memory"]
+
+    @abstractmethod
+    def _process(self, point: Tuple[float, float]) -> float:
+        """Defines how the IQ point will be processed.
+
+        Args:
+            point: An IQ point as a tuple of two float, i.e. (real, imaginary).
+
+        Returns:
+            Processed IQ point.
+        """
+
+    def process(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Modifies the data inplace by taking the real part of the memory and
+        scaling it by the given factor.
+
+        Args:
+            data: The data dict. IQ data is stored under memory.
+
+        Returns:
+            processed data: A dict with the data.
+        """
+
+        # Single shot data
+        if isinstance(data["memory"][0][0], list):
+            new_mem = []
+            for shot in data["memory"]:
+                new_mem.append([self.scale * self._process(iq_point) for iq_point in shot])
+
+            if self.average:
+                new_mem = list(np.mean(np.array(new_mem), axis=0))
+
+        # Averaged data
+        else:
+            new_mem = [self.scale * self._process(iq_point) for iq_point in data["memory"]]
+
+        return {self.__node_output__: new_mem}
```

Comment:
All of the current `DataAction`s return a dict with their `__node_output__`'s as the key and the result as the value. I wonder if this is better than just returning the data directly? It makes it possible for a DataAction to look at the keys in its input and behave differently depending on what is there. On the other hand, it makes it harder to swap in a different compatible `DataAction` because the following `DataAction` needs to know the key in advance. For example, to swap `ToReal` for `ToImag` the following `DataAction` would need to be able to accept `memory_real` and `memory_imag`. Also, I wonder about how much flexibility to allow in the input to a `DataAction` versus making separate `DataAction`s for the different cases.

Reply:
This is a good point. Returning the data directly will make it easier to swap in and out different DataActions in the chain; a highly desirable feature. What I like about the current structure is that it makes it easy to debug the data processing when it goes wrong. That being said, I think the current implementation of the history allows us to do just that. Perhaps the following change is the right way to go:
* return the data directly
* have each node check the format of the data, e.g. if a node expects single-shot IQ data then it should get a 3D list.

Reply:
Yes, that’s really good suggestion. Also regarding the comment below, the data attribute should be attached to data itself, not to DataAction. I would define dataclass to represent processed data, in which processing_level, ie discriminated, kerneled, ..., and return_type, ie avg/single, are stored along with the data. I don’t think DataAction level validation is not necessary, because most of users will use well designed default processing passes.

Reply:
I like this idea. If you did this nodes wouldn't work with dicts (unless the input data they are defined on is a dict) they would be `node(input_type) -> output_type` which makes them more like a regular function with extra metadata about types they accept.

Maybe one way of doing this would be when you define a node, you can haves abstract method you implement work on raw data, and optionally initialize it with a key value for looking up the data in an input result dict (maybe eventually result container class like Naoki suggests to make type checking easier).

This might look something like
```python
class DataAction(metaclass=ABCMeta):
    def __init__(self, data_key: Optional[str] = None):
        # Optional key to look up data value if input is dict
        self._data_key = data_key
    @abstractmethod
    def _process(self, data: any) -> any:
        pass

    def __call__(self, data) -> any:
        if isinstance(data, CircuitResult):
           if self._data_key is None or not hasattr(data, self._data_key):
                # raise error
           data = getattr(data, self._data_key)
        return self._process(data)
```
For now `CircuitResult` is `dict`, but that will lead to problems if the actual data type is a dict (like counts), so may be argument for a basic wrapper class to identity a result object like Naoki suggested.

If you did something like this these nodes could be:
```python
class IQPart(DataAction):
    def __init__(self, scale: Optional[float] = 1.0, average: bool = False,
                        data_key: str = "memory"):
        ...
        super().__init__(data_key)
```

Reply:
Anothing idea: If nodes don't typically have state, they are basically a wrapper for a function you could make a function wrapper class (or just build this into the base DataAction class) like:

```python
class DataActionFn(DataAction):
    def __init__(self, func: Callable, data_key: Optional[str] = None):
        # Optional key to look up data value if input is dict
        self._processing_func = func
        self._data_key = data_key

    def _process(self, data):
        return self._processing_func(data)
```

Then you could build up a DataProcessor by just passing in regular lambda functions if you wanted where you have in DataProcessor
```python
def append(self, node: Union[DataAction, Callable]):
    if not isinstance(node, DataAction):
        # Node is a regular function or some other callable
        node = DataActionFn(node)
    ...
```

Reply:
One use case I am wondering about is how a user would change the data processing at a lower level than what the experiment is targeted at. For example, the T1 experiment PR looks at `counts`. If I have my own kernel that I want to use, then I pass `meas_level=0` to `T1Experiment.run()` and pass my `DataProcessor` that takes level 0 data and produces counts (this support does not exist yet but probably this is passed to `run()`? I am bit uncertain about what goes in `run()` vs. `__init__()`). This seems okay, but it means that I need to know which level each experiment expects and pass it the processor to get it to that level. It would be nice if it were possible to pass all experiments the same object that knew how to go from level 0 to level 1 to level 2 (and average or not) and let the experiment decide which processor it needed to run for its analysis.

It might not be worth worrying about this too much in practice because with near-time compute I imagine there will be a way to set your own kernel in the backend, so the data processing does not have to be done by the client, but maybe the ability to do the processing on the client should still be supported.

Reply:
I think it still makes sense to support the data processing on the client side as this is how new/better/faster readout methods can be created.

Reply:
The changes in a0d0903 move us way from the intput/output key methodology and data is now directly returned.